[chip-tool] Add delay commands module to chip-tool with busy-wait com…

vivien-apple opened the pull request:
…mand

#### Problem

[hint] This is different than #23886 in the sense that it can be used at any time during an interactive session.
That makes it useful to try to not asked messages from the server, such as the messages from a subscription.

One of the thing this PR tries to achieved is to trigger an `INVALID_SUBSCRIPTION` status response when a subscription is in place. The mechanism used is to prevent any message processing from the stack for a duration longer than the subscription maximum interval + the mrp retries.

As an example:
```
$ chip-tool interactive start

# Once in interactive mode
>>> onoff subscribe on-off 1 5 0x12344321 1
>>> delay busy-wait 10000
```

The messages are queued during this interval (and not dropped) and so they are delivered to the message processing code afterwards. Which in turn trigger the stack to ack the reports but the subscription is closed on the other side.

andy31415 commented on examples/chip-tool/commands/delay/BusyWaitCommand.cpp:
```diff
@@ -0,0 +1,33 @@
+    auto start   = clock.GetMonotonicTimestamp();
+    chip::System::Clock::Milliseconds32 durationInMs(mDurationInMs);
+    while (clock.GetMonotonicTimestamp() - start < durationInMs)
+    {
```
how about at least a delayMS of 1 ms here? full 100% CPU busy look sounds really painful.

Also this really needs a comment as to why this is done in such a way or what uses busy wait has in general (so I can review it without looking at the PR history)

bzbarsky-apple replied:
@andy31415 A `delayMS` would be async and would not block message processing.  The point of this command is as a fault-injection tool: it blocks the event loop for some period of time, causing timeouts on the other side of the communication.

That said, we may be able to just `sleep()` or something instead of busy-waiting.  But I recall something failing when I tried that in the existing places where we have this sort of busy-wait fault-injection....

vivien-apple replied:
As Boris pointed out, the goal of this command is to block event loop processing, which is why is should not be async.

But this is right that I have not tried sleeping, as I remember it was not working too. Indeed, it does work for my purpose and I don't need to consume the whole CPU.

andy31415 replied:
Please add a comment here saying sleep using std is on purpose to block the main loop processing as a fault injection tool.

bzbarsky-apple replied:
This is all done now.

andy31415 commented:
@vivien-apple please add more details to the summary: why is this needed. The "hint" text does not seem to help me understand ... why is a busy loop better for interactive work?

andy31415 reviewed (dismissed):
The PR summary describes what is done (add a busy wait loop) but not **why**.

Please exaplain why a bit clearer. My instinct is to complain on any busy loop in our code - using 100% CPU to wait for a time seems wrong. It needs very strong justification in case this is the only thing possible.

vivien-apple commented:
@andy31415 I have updated the PR summary. 
About adding a comment, I have added one here already: https://github.com/project-chip/connectedhomeip/pull/24063/files#diff-cc6e5d55042827cc59f49f84ad79ff9ec700491da1a5f42355269c4ad7c26f0eR23
Is it not enough or is it misplaced such that it makes it hard to find ?

bzbarsky-apple commented on examples/chip-tool/commands/delay/BusyWaitCommand.h:
```diff
@@ -0,0 +1,55 @@
+#include "../common/CHIPCommand.h"
+
+/**
+ * This command block the event loop processing for a given amount of time.
```
```suggestion
 * This command blocks the event loop processing for a given amount of time.
```

bzbarsky-apple commented on examples/chip-tool/commands/delay/BusyWaitCommand.h:
```diff
@@ -0,0 +1,55 @@
+ *
+ */
+
+class BusyWaitCommand : public CHIPCommand
```
```suggestion
class BlockEventProcessingCommand : public CHIPCommand
```
perhaps?

vivien-apple replied:
I changed it for `SleepCommand`.

bzbarsky-apple commented on examples/chip-tool/commands/delay/BusyWaitCommand.h:
```diff
@@ -0,0 +1,55 @@
+class BusyWaitCommand : public CHIPCommand
+{
+public:
+    BusyWaitCommand(CredentialIssuerCommands * credIssuerCommands) : CHIPCommand("busy-wait", credIssuerCommands)
```
And here, replace busy-wait with something else?

Maybe just call this SleepCommand and use "sleep" for the actual command invocation?

vivien-apple replied:
I changed it for `SleepCommand`.
